### PR TITLE
Fixed #30439 - Added support for different plurals for the same language.

### PR DESCRIPTION
**Pull request by claudep**
ticket-30439

### CODE REVIEW

**Comment by claudep** [comment]
@uri-rodberg Any chance to test this code?

**Comment by uri-rodberg** [comment]
> @uri-rodberg Any chance to test this code?

@claudep It seems that this PR fixes what is not working for [Speedy Net](https://github.com/speedy-net/speedy-net) in Django 2.2:

[Speedy Net with Django==2.2.9](https://travis-ci.org/speedy-net/speedy-net/builds/638881102)
[Speedy Net with Django==2.2.9 + This PR included](https://travis-ci.org/speedy-net/speedy-net/builds/638881106)

I still have to check more, but it seems that everything in Speedy Net works with this PR.

If this PR is accepted, will it be possible to apply it to Django 2.2 as well?

File: django/utils/translation/trans_real.py [hunk 1]
```diff
@@ -58,6 +58,58 @@ def reset_cache(**kwargs):
...
+    """
+    Simulate a dict for DjangoTranslation._catalog so as multiple catalogs
+    with different plural equations are kept separate.
+    """
+    def __init__(self, trans=None):
+        self._catalogs = [trans._catalog.copy()] if trans else [{}]
+        self._plurals = [trans.plural] if trans else [lambda n: int(n != 1)]
+
+    def __getitem__(self, key):
+        return self._catalogs[0][key]
```
Comment (carltongibson) [inline django/utils/translation/trans_real.py]: Question: Why does this only pull from the first catalog?

Comment (claudep) [reply django/utils/translation/trans_real.py]: Good question, probably an oversight from my first POC trial.

**Comment by claudep** [comment]
Thanks for reviewing! As for more testing, I would much appreciate a review by @nijel, the initial reporter and an i18n expert.
And of course, the note about "Plural form and po files" definitely needs to be updated too! A release note would also be needed.

File: django/utils/translation/trans_real.py [hunk 2]
```diff
@@ -58,6 +58,63 @@ def reset_cache(**kwargs):
...
+        self._catalogs = [trans._catalog.copy()] if trans else [{}]
+        self._plurals = [trans.plural] if trans else [lambda n: int(n != 1)]
+
+    def __getitem__(self, key):
+        for cat in self._catalogs:
+            try:
+                return cat[key]
+            except KeyError:
+                pass
+        raise KeyError
```
Comment (m-aciek) [inline django/utils/translation/trans_real.py]: What would you say for:
```suggestion
        raise KeyError(key)
```

? That would make the error a bit more descriptive.

File: django/utils/translation/trans_real.py [hunk 3]
```diff
@@ -58,6 +58,63 @@ def reset_cache(**kwargs):
...
+            yield from cat.items()
+
+    def keys(self):
+        for cat in self._catalogs:
+            yield from cat.keys()
+
+    def update(self, trans):
+        # Merge if plural function is the same, else prepend.
+        for cat, plural in zip(self._catalogs, self._plurals):
+            if trans.plural.__code__ == self._plurals[0].__code__:
```
Comment (m-aciek) [inline django/utils/translation/trans_real.py]: Didn't you mean:
```suggestion
            if trans.plural.__code__ == plural.__code__:
```
here?

Comment (claudep) [reply django/utils/translation/trans_real.py]: Sure, thanks.

File: django/utils/translation/trans_real.py (hunk 2 above)
Comment (claudep) [reply django/utils/translation/trans_real.py]: Good idea.

**Comment by nijel** [comment]
I've finally got to testing this and I can confirm it works fine for Weblate. Looking at the code it also looks good. Thanks!

Has this chance to land in 2.2?

**Comment by uri-rodberg** [comment]
> @uri-rodberg Any chance to test this code?

@claudep Do we need to test Speedy Net again? I see you changed this PR after my test.

**Comment by claudep** [comment]
> @claudep Do we need to test Speedy Net again? I see you changed this PR after my test.

No, it was a minor thing, no need to re-test.

> Has this chance to land in 2.2?

I'm afraid not.

**Comment by uri-rodberg** [comment]
Any update for this PR? @carltongibson

File: docs/releases/2.2.12.txt [hunk 4]
```diff
@@ -9,4 +9,5 @@ Django 2.2.12 fixes several bugs in 2.2.11.
 Bugfixes
 ========
 
-* ...
+* Added the ability to handle po files containing different plural equations
```
Comment (felixxm) [inline docs/releases/2.2.12.txt]: `po` -> ``` ``.po`` ```

File: docs/topics/i18n/translation.txt [hunk 5]
```diff
@@ -277,14 +277,9 @@ In a case like this, consider something like the following::
...
-
-    Django does not support custom plural equations in po files. As all
-    translation catalogs are merged, only the plural form for the main Django po
-    file (in ``django/conf/locale/<lang_code>/LC_MESSAGES/django.po``) is
-    considered. Plural forms in all other po files are ignored. Therefore, you
-    should not use different plural equations in your project or application po
-    files.
+.. versionchanged: 2.2.12
+
+    Added support for different plural equations in po files.
```
Comment (felixxm) [inline docs/topics/i18n/translation.txt]: `po` -> ``` ``.po`` ```

File: docs/releases/2.2.12.txt [hunk 6]
```diff
@@ -9,4 +9,5 @@ Django 2.2.12 fixes several bugs in 2.2.11.
 Bugfixes
 ========
 
-* ...
+* Added the ability to handle ``.po`` files containing different plural
```
Comment (felixxm) [inline docs/releases/2.2.12.txt]: Maybe: `Django 2.2.12 fixes a bug in 2.2.`, I don't think that we will decide to backport more bugfixes.

Comment (carltongibson) [reply docs/releases/2.2.12.txt]: OK, updated. 👍

**Comment by uri-rodberg** [comment]
Thank you for accepting this PR and applying it to Django 2.2 and 3.0 too.

**Comment by uri-rodberg** [comment]
Hi, I just want to say that I upgraded Django to 2.2.12 in [Speedy Net](https://github.com/speedy-net/speedy-net) and then upgraded again to 3.0.5, and everything works as expected. I didn't change anything in our translation files and they are still using the old plural forms formula like before (`"Plural-Forms: nplurals=2; plural=(n != 1);\n"`).

**Comment by claudep** [comment]
Thanks for the feedback!

**Pull request merged**